flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/interval/asyncprocess/AsyncTimeIntervalJoin.java

```diff
@@ -0,0 +1,347 @@
+                                rightExpirationTime =
+                                        calExpirationTime(leftOperatorTime, rightRelativeSize);
+                                removeExpiredRows(
+                                        rightExpirationTime,
+                                        rightCache,
+                                        leftTimerState,
+                                        ctx,
+                                        false);
+                            }
+                        });
+
+        rightTimerState
+                .asyncValue()
+                .thenAccept(
+                        rightCleanUpTime -> {
+                            if (rightCleanUpTime != null && timestamp == rightCleanUpTime) {
+                                leftExpirationTime =
+                                        calExpirationTime(rightOperatorTime, leftRelativeSize);
+                                removeExpiredRows(
+                                        leftExpirationTime, leftCache, rightTimerState, ctx, true);
+                            }
+                        });
+    }
+
+    @Override
+    protected boolean iterateRightCache(
+            RowData leftRow, long rightQualifiedLowerBound, long rightQualifiedUpperBound)
+            throws Exception {
+        Set<Long> toRemove = new HashSet<>();
+        final AtomicBoolean emitted = new AtomicBoolean(false);
+        CompletableFuture<Boolean> emitResult = new CompletableFuture<>();
+
+        Iterator<Map.Entry<Long, List<Tuple2<RowData, Boolean>>>> rightIterator =
+                rightCache.iterator();
+        rightCache
+                .asyncEntries()
+                .thenCompose(
+                        iter ->
+                                iter.onNext(
+                                        entry -> {
+                                            boolean entryEmitted =
+                                                    tryMatchLeftRowsByTime(
+                                                            leftRow,
+                                                            rightIterator.next(),
+                                                            rightQualifiedLowerBound,
+                                                            rightQualifiedUpperBound,
+                                                            toRemove);
+                                            if (entryEmitted) {
+                                                emitted.set(true);
+                                            }
+                                        }))
+                .thenAccept(
+                        V -> {
+                            emitResult.complete(emitted.get());
+                            for (Long key : toRemove) {
+                                rightCache.asyncRemove(key);
+                            }
+                        });
+        return emitResult.get();
```

Comment:
We better not blocking `get()` here. It will block the task thread as well as those callbacks for async state requests.

Reply:
👍I changed the return value to `StateFuture<Boolean>`.